Fix standard deviation overflow(#95)

y-okumura-isp opened the pull request:
Since I haven't written unit test, l paste the example_loop and numpy results instead.
I checked in 64 bit(ubuntu x86_64) and 32 bit(Raspberry Pi B+), pasted results are in 32bit.

Standard deviations are not equal exactly because it's a floating point calculation.

```
# ./example_loop -s rr -i 10000 -d 200mb -f out.txt
Writing results to file: out.txt
Initial major pagefaults: 0
Initial minor pagefaults: 51508
rttest statistics for out.txt:
  - Minor pagefaults: 0
  - Major pagefaults: 0
  Latency (time after deadline was missed):
    - Min: 8302 ns
    - Max: 135160 ns
    - Mean: 11098.962400 ns
    - Standard deviation: 2934.104804

# python
>>> import numpy as np
>>> np.loadtxt("out.txt", skiprows=1)[:, 2].std()
2934.929376524457
```

dirk-thomas commented:
@y-okumura-isp please update the PR title and use a descriptive text of the actual change. The reference to the issue this addresses should then go into the description with a proper keyword so GitHub closes the referenced issue automatically when this PR is merged.

y-okumura-isp commented:
@clalancette Thank you for comment.

> Please add a test when you have some time, that way we can ensure we don't break it in the future.

I did in 3be4434.
There are many changes, so let me explain.

(1) math_utils.hpp, test_math_utils.cpp
  - As calculation routine is embedded in `Rttest::calculate_statistics`, so I needed to split this as a function and write test.
    As utils.h is C-style, I added math_utils.hpp (although my PR 96 changes utils.h to utils.hpp...) 
  - test_api.cpp seems to be test for rttest API, so I added test_math_utils.cpp.

(2) minnor fix
  - Type of `latency_diff` was `std::vector<int64_t>`, but `std::vector<double>` is preferable because average (not integer) is subtracted.

By the way as calcurating statistics is very common, I think it's better to use some statistics library or math library such as eigen.
But I didn't because changes will be bigger.

clalancette commented:
> By the way as calcurating statistics is very common, I think it's better to use some statistics library or math library such as eigen.
> But I didn't because changes will be bigger.

I agree with not bringing in another dependency.

clalancette commented on rttest/include/rttest/math_utils.hpp:
```diff
@@ -0,0 +1,49 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
```
```suggestion
// Copyright 2020 Open Source Robotics Foundation, Inc.
```

clalancette commented on rttest/include/rttest/math_utils.hpp:
```diff
@@ -0,0 +1,49 @@
+#ifndef RTTEST__MATH_UTILS_HPP_
+#define RTTEST__MATH_UTILS_HPP_
+
+#include <vector>
+#include <cmath>
+#include <algorithm>
+#include <numeric>
```
Nit: alphabetize this list.

y-okumura-isp replied:
6993c6f

clalancette commented on rttest/include/rttest/math_utils.hpp:
```diff
@@ -0,0 +1,49 @@
+    vec.begin(), vec.end(), diff.begin(),
+    [mean](auto x) -> double {return x - mean;});
+
+  // forst divide by sqrt(n)
```
```suggestion
  // first divide by sqrt(n)
```

clalancette commented on rttest/test/test_math_util.cpp:
```diff
@@ -0,0 +1,26 @@
+
+  std::vector<double> v_double = {1.1, 2.2, 3.3, 4.9};
+  EXPECT_NEAR(1.404, calculate_stddev(v_double), 0.001);
+}
```
Can you please add a test here that shows the problem from #95?

y-okumura-isp replied:
bcf95fe.

y-okumura-isp commented:
Thank you for many comments.
I've done with all comments(if no oversight).

clalancette commented:
Thanks for the contribution; this looks good to me now.  I'm going to hold off on running CI and merging this until after Foxy is released; please feel free to ping me after June 8th if I forget about this.

clalancette commented:
@y-okumura-isp Oh, one other thing; can you please rebase and add a "Signed-off-by" line to all of your commits?  That way our DCO check will pass.  Thanks.

y-okumura-isp commented:
> can you please rebase and add a "Signed-off-by" line to all of your commits?

Sure, I fixed.
Thank you for reviewing so patiently.

clalancette commented:
> Thank you for reviewing so patiently.

You are welcome, thanks for iterating with us.

clalancette commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=11038)](http://ci.ros2.org/job/ci_linux/11038/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=6363)](http://ci.ros2.org/job/ci_linux-aarch64/6363/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8988)](http://ci.ros2.org/job/ci_osx/8988/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10959)](http://ci.ros2.org/job/ci_windows/10959/)

clalancette commented:
Windows is failing CI, but that's because of infrastructure issues.  Since this package doesn't build on Windows *anyway*, I'm going to go ahead and merge without that.  Thanks for the contribution!
